File: backend/apps/api/envelope.py
"""Unified API response envelope — Sprint 1 Track 3."""
from __future__ import annotations

from typing import Any

from apps.core.logging import get_trace_id


def success_envelope(
    data: Any = None,
    *,
    message: str = "",
    meta: dict[str, Any] | None = None,
) -> dict[str, Any]:
    merged_meta = {"trace_id": get_trace_id()}
    if meta:
        merged_meta.update(meta)
    return {
        "success": True,
        "data": data,
        "message": message or "",
        "errors": [],
        "meta": merged_meta,
    }


def error_envelope(
    message: str,
    *,
    code: str = "api_error",
    errors: list[dict[str, Any]] | None = None,
    meta: dict[str, Any] | None = None,
) -> dict[str, Any]:
    merged_meta = {"trace_id": get_trace_id()}
    if meta:
        merged_meta.update(meta)
    detail_errors = errors or [{"code": code, "message": message}]
    return {
        "success": False,
        "data": None,
        "message": message,
        "errors": detail_errors,
        "meta": merged_meta,
    }


def is_envelope(payload: Any) -> bool:
    return isinstance(payload, dict) and "success" in payload and "data" in payload

File: frontend/src/platform/envelope.ts
/**
 * Unified API response envelope — Track 2 client normalization.
 * Track 3 will make Django emit this natively.
 */

export interface ApiEnvelope<T = unknown> {
  success: boolean;
  data: T | null;
  message: string;
  errors: Array<{ code?: string; message: string; field?: string }>;
  meta: Record<string, unknown>;
}

function isEnvelope(value: unknown): value is ApiEnvelope {
  return (
    !!value &&
    typeof value === "object" &&
    "success" in value &&
    typeof (value as ApiEnvelope).success === "boolean"
  );
}

/** Normalize Django raw success OR error OR future envelope into ApiEnvelope. */
export function normalizeResponse<T = unknown>(
  body: unknown,
  httpOk: boolean,
  status: number
): ApiEnvelope<T> {
  if (isEnvelope(body)) {
    return body as ApiEnvelope<T>;
  }

  if (
    body &&
    typeof body === "object" &&
    "error" in body &&
    (body as { error: unknown }).error &&
    typeof (body as { error: unknown }).error === "object"
  ) {
    const err = (body as { error: { code?: string; message?: unknown; trace_id?: string } }).error;
    const message =
      typeof err.message === "string"
        ? err.message
        : err.message != null
          ? JSON.stringify(err.message)
          : "Request failed";
    return {
      success: false,
      data: null,
      message,
      errors: [{ code: err.code ?? `http_${status}`, message }],
      meta: err.trace_id ? { trace_id: err.trace_id } : {},
    };
  }

  if (!httpOk) {
    return {
      success: false,
      data: null,
      message: `HTTP ${status}`,
      errors: [{ code: `http_${status}`, message: `HTTP ${status}` }],
      meta: {},
    };
  }

  return {
    success: true,
    data: body as T,
    message: "",
    errors: [],
    meta: {},
  };
}
